src/Function.cpp

```diff
@@ -77,6 +77,53 @@ static long cur_func_idx;				// Index into FuncList that we are currently workin
 static bool param_first=true;			// Flag to track output of commas
 static int builtin_functions_cnt;
 
+void
+FunctionAttribute::OutputAttribute(std::ostream &out)
+{
+	out << " __attribute__((" << attribute << "))";
+}
+
+BooleanFunctionAttribute::BooleanFunctionAttribute(int prob, string name)
+{
+	is_attribute = prob;
+	attribute = name;
+}
+
+void
+BooleanFunctionAttribute::OutputAttributes(std::ostream &out)
+{
+	if(rnd_flipcoin(is_attribute))
+		OutputAttribute(out);
+}
+
+MultiValuedAttribute::MultiValuedAttribute(int prob, string name, vector<string> arguments)
+{
+	is_attribute = prob;
+	attribute = name;
+	attribute_values = arguments;
+}
+
+void
+MultiValuedAttribute::OutputAttributes(std::ostream &out)
+{
+	if(rnd_flipcoin(is_attribute)){
+		attribute += attribute_values[rnd_upto(attribute_values.size())];
```

Comment:
Please verify that it's not leading to an off-by-one error. I don't like how you append to `attribute`. Please use second argument to `OutputAttributes`.

Reply:
Passing above appended string as extra parameter to `OutputAttributes`